solutions/cfgym-101102D.mdx

```diff
@@ -0,0 +1,259 @@
+---
+id: cfgym-101102D
+source: CF
+title: Rectangles
+author: Kevin Sheng
+---
+
+Let's first simplify this problem, and then apply the solution of that problem to this one.
+
+Say we had a matrix of all `0`'s and `1`'s. How many submatrices of all `1`'s are there?
+
+We first define a 2D array, $\texttt{longest\_consec}$, that contains for each cell how many `1`'s there are to the right.  
+For example, this matrix:
+
+<table>
+	<tr>
+		<td>1</td>
+		<td>0</td>
+		<td>1</td>
+		<td>1</td>
+	</tr>
+	<tr>
+		<td>0</td>
+		<td>1</td>
+		<td>1</td>
+		<td>1</td>
+	</tr>
+	<tr>
+		<td>0</td>
+		<td>0</td>
+		<td>0</td>
+		<td>1</td>
+	</tr>
+</table>
+
+Would have the following $\texttt{longest\_consec}$:
+
+<table>
+	<tr>
+		<td>1</td>
+		<td>0</td>
+		<td>1</td>
+		<td>2</td>
+	</tr>
+	<tr>
+		<td>0</td>
+		<td>1</td>
+		<td>2</td>
+		<td>3</td>
+	</tr>
+	<tr>
+		<td>0</td>
+		<td>0</td>
+		<td>0</td>
+		<td>1</td>
+	</tr>
+</table>
+
+Now, we go through each column and then for each row in reverse,
+finding out for each cell the number of submatrices of all `1`'s
+that have said cell as an upper left corner.
+
+How do we do this?  
+First, let's initialize a stack $\texttt{rel\_streaks}$ and a number $\texttt{curr\_sum}$.  
+$\texttt{rel\_streaks}$ will have elements each with two things:
+the value of $\texttt{longest\_consec}$ for that element and the difference in row number
+(aka distance) between that element and the next element in the stack.  
+$\texttt{curr\_sum}$, after processing, will contain
```

Comment:
Optional stuff, I'm just being annoying
```
First, let's initialize a stack $\texttt{rel\_streaks}$ which will have elements with two properties: the value of $\texttt{longest\_consec}$ for that element and the difference in row number (aka distance) between that element and the next element in the stack. 

Also initialize a number $\texttt{curr\_sum}$, which after processing, will contain
the number of submatrices of all `1`'s that start at each cell.
```
Seems more intuitive to talk about the stack at the top, and then about curr_sum at the bottom.

Reply:
yeah that makes sense